### PR TITLE
Fixes to evaluator and various compilers

### PR DESCRIPTION
**Pull request by bksaiki**
- Fixes issues #85, #86, and #87 
- Fixes casting and precision changes in the imperative compiler
- Fixes/cleans up TeX output (Herbie)
- Minor changes to Makefile

### CODE REVIEW

File: src/imperative.rkt [hunk 1]
```diff
@@ -244,7 +244,8 @@
        [(and (or (equal? (language-name (*lang*)) "c")  ; Only C and Scala need to emit a temporary variable
                  (equal? (language-name (*lang*)) "scala"))
              (or (not (equal? curr-prec new-prec))
-                 (not (equal? curr-round new-round))))
+                 (not (equal? curr-round new-round)))
+             (list? body))                            ; except if it is a no op
```
Comment (monadius) [inline src/imperative.rkt:248]: This change does not fix #86. Consider the following core:
```
(FPCore simple (x)
  :precision binary64
  (! :precision binary32 (! :precision binary32 x)))
```
The result is the same as before:
```c
double ex0(double x) {
	float tmp = x;
	return (double) tmp;
}
```

File: src/imperative.rkt [hunk 2]
```diff
@@ -279,7 +280,7 @@
     (define ctx (ctx-reserve-names default-ctx (*reserved-names*)))
 
     (define func-name 
-      (let-values ([(cx fname) (ctx-unique-name ctx (string->symbol name))])
+      (let-values ([(cx fname) (ctx-unique-name ctx (string->symbol (fix-name name)))])
```
Comment (monadius) [inline src/imperative.rkt:283]: Why did you add `fix-name` here?

Comment (bksaiki) [reply src/imperative.rkt:283]: Oops. Forgot to remove that. That was part of a change I reverted.

File: src/imperative.rkt (hunk 1 above)
Comment (bksaiki) [reply src/imperative.rkt:248]: Ahh thanks. I guess this one is a little more complicated.

Comment (monadius) [reply src/imperative.rkt:248]: I think the following idea will work here. `convert-expr` should return both the result expression (as a string) and its type (precision). Then types of all temporary variables are defined by types returned by `convert-expr`.

Comment (bksaiki) [reply src/imperative.rkt:248]: That's actually what the `smt` compiler does. The simpler solution would be to also check for `!` in the body of the expression.

Comment (bksaiki) [reply src/imperative.rkt:248]: Now that I think about it, constants need to be rounded, so a temporary variable should be emitted unless the body contains another annotation or a variable.

Comment (monadius) [reply src/imperative.rkt:248]: It is hard to check all possible cases. For example:
```
(FPCore simple (x)
  :precision binary64
  (! :precision binary32 (let ((y x)) y)))
```
So it would be better to make changes similar to the `smt` compiler. Then it will also be possible to reuse `imperative` functions in the `smt` compiler (`convert-application` should be generalized to do this).

Comment (bksaiki) [reply src/imperative.rkt:248]: That's a good counterexample. Ok, I'll try to get this fixed quickly.

File: src/imperative.rkt [hunk 3]
```diff
@@ -245,27 +260,44 @@
...
-    [(list digits m e b) (convert-constant ctx (digits->number m e b))]
-    [(? constant?) (convert-constant ctx expr)]
-    [(or (? number?) (? hex?)) (convert-constant ctx expr)]
-    [(? symbol?) (ctx-lookup-name ctx expr)]))
+      (define-values (args* precs*)
+        (for/lists (args* precs*) ([arg args])
+          (convert-expr arg #:ctx ctx #:indent indent)))
+      (define cast? ; C implicit casting: check if arguments need to be explicitly casted
+        (and (equal? (language-name (*lang*)) "c")
+             (andmap (λ (x) (positive? (cmp-prec (ctx-lookup-prop ctx ':precision) x))) precs*)))
```
Comment (monadius) [inline src/imperative.rkt:285]: Shouldn't it be the other way around, i.e., `negative?`?

This core
```
(FPCore exp32 (x)
  :precision binary64
  (! :precision binary32 (exp x)))
```
is translated to:
```c
double ex3(double x) {
	float tmp = expf(x);
	return tmp;
}
```

And this core
```
(FPCore exp64 (x)
  :precision binary32
  (! :precision binary64 (exp x)))
```
is translated to:
```c
float ex4(float x) {
	double tmp = exp((double) x);
	return tmp;
}
```
I think it makes more sense to convert `double` to `float` explicitly rather than `float` to `double`.

In general, I am not sure what is the right behavior in situations like this. The core `exp32` cannot be faithfully translated into C code. Maybe it is better to report errors for such cores.

File: src/imperative.rkt [hunk 4]
```diff
@@ -98,142 +98,157 @@
...
-              indent
-              (format-cond (trim-infix-parens (convert-expr cond #:ctx ctx #:indent indent))))
-            (printf "~a\t~a\n" indent
-                    (convert-assignment outvar
-                                        (convert-expr elif #:ctx ctx
-                                                      #:indent (format "~a\t" indent))))])))
-      outvar]
+              indent (format-cond (trim-infix-parens cond*)))
+            (printf "~a\t~a\n" indent (convert-assignment (car out) elif*))
+            (loop (cdr branches) first? out)])))]
```
Comment (monadius) [inline src/imperative.rkt:175]: `if` is broken. Consider the following example:
```
(FPCore if (x)
  (if TRUE x (let ([y (/ 1 0)]) y)))
```
The translated result is:
```c
double ex6(double x) {
	double tmp;
	if ((double) TRUE) {
		tmp = x;
		double y = 1.0 / 0.0;
	} else {
		tmp = y;
	}
	return tmp;
}
```
Maybe it is possible to fix this issue by changing `current-output-port` for `ift` and `elif` expressions.

File: src/imperative.rkt (hunk 3 above)
Comment (bksaiki) [reply src/imperative.rkt:285]: Oops, totally redundant. The reason for explicit casting was this example:
```
(FPCore add2 (x y)
(FPCore (x y)
  :precision binary64 
  (- (! :precision binary32 (+ x y)) 
     (! :precision binary32 (- x y))))
```
Without the cast the final type would be a `float`.

File: src/imperative.rkt (hunk 4 above)
Comment (bksaiki) [reply src/imperative.rkt:175]: Thanks. I just messed up the order of the `printf`'s.

File: src/imperative.rkt [hunk 5]
```diff
@@ -150,27 +150,31 @@
...
             (values (car out) (cdr out))]
            [(list _ cond elif) ; otherwise => else if branch
             (define-values (cond* cond-prec) (convert-expr cond #:ctx ctx #:indent indent))
-            (define-values (elif* elif-prec) (convert-expr elif #:ctx ctx #:indent (format "~a\t" indent)))
             (printf
               (if (equal? (language-name (*lang*)) "sollya") ; Sollya has slightly different else if
                   "~a} else if ~a then {\n"
                   "~a} else if ~a {\n")
               indent (format-cond (trim-infix-parens cond*)))
+            (define-values (elif* elif-prec) (convert-expr elif #:ctx ctx #:indent (format "~a\t" indent)))
```
Comment (monadius) [inline src/imperative.rkt:177]: `else if` contains a bug:
```
(FPCore if-else-cond (x)
  (if TRUE x (if (let ([t FALSE]) t) x x)))
```
```c
double ex9(double x) {
	double tmp;
	if (TRUE) {
		tmp = x;
	double t = FALSE;
	} else if (t) {
		tmp = x;
	} else {
		tmp = x;
	}
	return tmp;
}
```
This bug can be fixed by removing `collect-branches` and by translating all `if` conditions separately.

Comment (bksaiki) [reply src/imperative.rkt:177]: Unfortunately, the `collect-branches` has to stay; the feature was imported from another project. I think the better fix is to convert all of the conditions before the first `if`. That way, the conditions passed to the loop are either expressions or variables, e.g. `TRUE` or `t`.

Comment (pavpanchekha) [reply src/imperative.rkt]: Woah good bug! Yeah I think Brett's idea will work. It does mean the C code might do some more work than the FPBench, but you can always fix that by inlining the lets...

**Review by pavpanchekha — APPROVED** [review]
I'm not sure if every issue @monadius had has been resolved, but I looked over the code and it looks good. I'm glad all these tests are being merged as regression cases.

**Comment by monadius** [comment]
@pavpanchekha I think it is OK to merge this pull request. There are still some issues with examples like:
```
(FPCore if-else-cond (x)
  (if TRUE x (if (let ([t (/ 1 0)]) t) x x)))
```
(1 / 0 may raise an exception in some languages).

More importantly, some tests fail now. Do you plan to address this issue in a separate pull request?

**Comment by bksaiki** [comment]
@monadius I'm planning on getting the other compilers working before I merge. The important issue was fixing the C compiler since it's arguably the most useful one. Plus, now that I have more cores for testing, it'll be much easier to fix up the other tools. 
Thanks for finding all these issues by the way. It was quite helpful. This PR might stay open for a few days, so if you notice anything else, please comment here.

As for your exception comment, I'm not sure how involved the compilers should be in checking for such problems. You previously brought up an FPCore program that can't be faithfully rounded, and I've though of a few more since then. I think there should definitely be a warning of some sort, so that needs to be added. 

Division by zero seem a little naive; such a program should crash(?) in an FPCore evaluator and in whatever language it gets converted to. However, I will note that the Sollya compiler translates `/` into a custom function that warns of division by zero since Sollya returns NaN, and the Daisy tool (see the Scala compiler), throws an error for `pow(b, x)` when `x` is negative (so maybe we should safeguard against that?).

**Comment by monadius** [comment]
@bksaiki I don't think that my example with division by zero should crash in any language including the FPCore evaluator. The else branch of this example should be evaluated only if the condition is `false` (but it is always `true`). In particular, the FPCore evaluator does not evaluate the else branch at all and always returns a correct result. I don't think it is possible to fix this issue without removing `collect-branches`. But it is possible to provide an option to turn `collect-branches` on and off.

I am also not sure how strict compilers should be with benchmarks which they cannot translate faithfully. One possible solution is to print a warning for such benchmarks and also add a new option to turn warnings into errors.

**Comment by monadius** [comment]
I don't understand the behavior of the following benchmark:
```
(FPCore main ()
  :precision binary32
  (let ([x 0x1p0]
        [y 0x1p-30])
    (! :precision binary64 (+ x y))))
```
Both evaluator and core2c return `1.0f` for this benchmark. I would expect to get `1 + 2^-30 = 1.0000000009313226`. Does the top level precision annotation imply an implicit cast to this precision? I cannot find anything about this implicit cast in the standard.

**Comment by bksaiki** [comment]
@monadius Oh, my bad. I was talking about division by zero issues in general. Sorry, for the confusion. Yes, I think you've found _another_ bug. I would use [Titanic](http://titanic.uwplse.org/) to evaluate FPCore programs. It's a much more sound evaluator. It returns 1.0000000009313226 for the example you provided.

**Comment by monadius** [comment]
My goal is to update the FPTaylor compiler and I want to be sure that I understand the standard correctly.

Here is another useful benchmark which tests the double rounding issue:
```
(FPCore no-double-rounding ()
    :precision binary64
    (let ([x 0x1.000002p-1]
          [y 0x1.ffffffffp-26])
        (! :precision binary32 (+ x y))))
```
The result of this core is different from
```
(FPCore double-rounding ()
    :precision binary64
    (let ([x 0x1.000002p-1]
          [y 0x1.ffffffffp-26])
        (! :precision binary32 (cast (! :precision binary64 (+ x y))))))
```

File: src/imperative.rkt [hunk 6]
```diff
@@ -98,142 +98,167 @@
...
-       (printf "~a}~a\n" indent (if (equal? (language-name (*lang*)) "sollya") ";" ""))
+        (let-values ([(val* _) (convert-expr update #:ctx ctx* #:indent indent*)])
+          (printf "~a\t~a\n" indent (convert-assignment var* val*))))
+
+      (define-values (cond** cond**-prec) (convert-expr cond #:ctx ctx* #:indent indent*))
+      (printf "~a\t~a\n" indent (convert-assignment test-var cond**))
+      (printf "~a}~a\n" indent (if (equal? (language-name (*lang*)) "sollya") ";" ""))
       (convert-expr retexpr #:ctx ctx* #:indent indent)]
 
     ; Ignore all casts
```
Comment (monadius) [inline src/imperative.rkt:258]: I think this comment ("Ignore all casts") should be removed because casts are not ignored. This comment should be removed from core2smtlib2.rkt also.

Comment (bksaiki) [reply src/imperative.rkt:258]: Oops. Those are obsolete comments.
